### PR TITLE
Save/Share trips

### PR DESCRIPTION
**Pull request by Cindyzhang977**
On save, create trip if trip doesn't exist, else update trip.

- Save button re-enabled when trip name is updated or route order is changed 
  - `tripObject` changed to state to keep `tripId` changes
- redirect to route/edit page from saved trips
  - changes in `ExploreView` reflect in `Routeview`, toggling between these two pages works
- compare attraction names in `ExploreView`, because `photoReference` somehow changes for the same attraction/image 
- share trip modal pops up on `SearchView` and `RouteView`
  - moved share trip modal to be it's own component

image

### CODE REVIEW

File: frontend/src/saved-trips/SavedTrip.js [hunk 1]
```diff
@@ -1,28 +1,86 @@
...
+import ShareTripModal from '../share-trip/ShareTripModal.js';
 
 /**
  * Creates single saved trip.
  */
-function SavedTrip({ name, id }) {
+function SavedTrip({ tripId }) {
+  const [showShareModal, setShowShareModal] = useState(false);
+
+  function setTripObjectDispatch(state, action) {
```
Comment (genedwards) [inline frontend/src/saved-trips/SavedTrip.js:17]: nit: setTripObjectReducer

File: frontend/src/saved-trips/SavedTrip.js [hunk 2]
```diff
@@ -1,28 +1,86 @@
...
+    return {
+      tripId: state.tripId,
+      attractions: parsedAttractions,
+      centerLocation: JSON.parse(action.centerLocation),
+      isOptimized: action.isOptimized,
+      searchText: JSON.parse(action.searchText),
+      tripName: JSON.parse(action.tripName),
+    };
+  }
+  const [tripObject, setTripObject] = useReducer(setTripObjectDispatch, {
```
Comment (genedwards) [inline frontend/src/saved-trips/SavedTrip.js:39]: nit: following ExploreView.js's pattern, `[tripObject, setTripObjectDispatch]`

File: frontend/src/saved-trips/SavedTrip.js [hunk 3]
```diff
@@ -1,28 +1,86 @@
...
+    tripId: tripId,
+    attractions: [],
+  });
+
+  useEffect(() => {
+    getTrip(tripId).then((trip) => setTripObject(trip));
```
Comment (genedwards) [inline frontend/src/saved-trips/SavedTrip.js:45]: would it make sense to JSON.parse within `getTrip` before it returns the trip object?

File: frontend/src/RouteView.js [hunk 4]
```diff
@@ -56,18 +51,21 @@ function RouteView({ loggedIn, userEmail }) {
...
+      console.log('create trip');
+      createTrip(userEmail, tripObject)
+        .then((res) => res.json())
+        .then((json) => setTripObject({ ...tripObject, tripId: json.tripId }));
     } else {
-      updateTrip(tripObject.id, tripObject);
+      console.log('update');
+      updateTrip(tripObject.tripId, tripObject);
     }
+    setIsSaved(true);
```
[the remark this answers is not in the document]
Comment (Cindyzhang977) [reply frontend/src/RouteView.js:68]: There isn't really a reason other than it makes sense to set saved to true after the calls are made to the endpoint (even though it's asynchronous).

File: frontend/src/routingUtils.js [hunk 5]
```diff
@@ -16,7 +16,6 @@ export const getQueryParameters = (query) => {
  * @param {object} attractions list of selected attractions
  */
 export const handleRouting = (history, page, tripObject, attractions) => {
```
[the remark this answers is not in the document]
Comment (Cindyzhang977) [reply frontend/src/routingUtils.js:18]: oops I accidentally removed this line `tripObject.attractions = attractions;`

File: frontend/src/saved-trips/SavedTrip.js [hunk 6]
```diff
@@ -1,28 +1,86 @@
...
+  function setTripObjectDispatch(state, action) {
+    const parsedAttractions = [];
+    const attractions = JSON.parse(action.attractions);
+    attractions.forEach((attr) => {
+      const parsedAttr = {
+        name: JSON.parse(attr.name),
+        photoUrl: JSON.parse(attr.photoUrl),
+        lat: JSON.parse(attr.lat),
+        lng: JSON.parse(attr.lng),
+        routeIndex: attr.routeIndex,
```
[the remark this answers is not in the document]
Comment (Cindyzhang977) [reply frontend/src/saved-trips/SavedTrip.js:26]: These have an extra set of quotes (""trip name""), and JSON.parse removes them.

File: frontend/src/tripUtils.js [hunk 7]
```diff
@@ -1,8 +1,10 @@
 import { fetchJson } from './fetchJson.js';
 
 export const createTrip = async (email, tripData) => {
-  const url = `/api/v1/createTrip?email=${email}&tripData=${JSON.stringify(tripData)}`;
-  await fetch(url, { method: 'POST' });
+  const url = `/api/v1/createTrip?email=${email}&tripData=${encodeURIComponent(
+    JSON.stringify(tripData)
+  )}`;
+  return await fetch(url, { method: 'GET' });
```
Comment (RCAVelez) [inline frontend/src/tripUtils.js:7]: fetch is get by default, you can remove this `{ method: 'GET' }`

File: frontend/src/ExploreView.js [hunk 8]
```diff
@@ -278,7 +290,7 @@ function Explore() {
    */
   function toggleSelection(targetAttraction) {
     const targetAttrIndexInSelected = selectedAttractions.findIndex(
-      (attraction) => attraction.photoUrl === targetAttraction.photoUrl
+      (attraction) => attraction.name === targetAttraction.name
```
Comment (RCAVelez) [inline frontend/src/ExploreView.js:293]: I'm wondering if uniqueness for the id of an attraction is greater within a photoUrl than from name. Are there any problems with this?

File: frontend/src/saved-trips/SavedTrip.js (hunk 2 above)
Comment (Cindyzhang977) [reply frontend/src/saved-trips/SavedTrip.js:39]: moved Json parsing to tripUtils

File: frontend/src/saved-trips/SavedTrip.js (hunk 1 above)
Comment (Cindyzhang977) [reply frontend/src/saved-trips/SavedTrip.js:17]: moved json parsing to trip utils

File: frontend/src/ExploreView.js (hunk 8 above)
Comment (Cindyzhang977) [reply frontend/src/ExploreView.js:293]: Yeah, the urls somehow are different even though the attraction is the same

File: frontend/src/RouteView.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (genedwards) [reply frontend/src/RouteView.js:66]: sounds good, no comment needed

File: frontend/src/tripUtils.js [hunk 9]
```diff
@@ -1,18 +1,40 @@
...
 
 export const getTrip = async (tripId) => {
   const url = `/api/v1/readTrip?tripId=${tripId}`;
-  return await fetchJson(url);
+  const json = await fetchJson(url);
+  const parsedAttractions = [];
+  const attractions = JSON.parse(json.attractions);
+  attractions.forEach((attr) => {
+    const parsedAttr = {
+      name: JSON.parse(attr.name),
```
Comment (genedwards) [inline frontend/src/tripUtils.js:17]: it still seems odd to me that we JSON.parse the attractions, and it's inner is still stringified, makes me think that we're doing something wrong when storing it
perhaps a separate issue

File: frontend/src/share-trip/ShareTripModal.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+ * @param {function} setShowShareModal toggle showShareModal
+ * @param {string} tripName the name of the trip to display in the modal
+ * @param {string} url the link to the route view of the trip to share
+ */
+function ShareTripModal({ showShareModal, setShowShareModal, tripName, url }) {
+  const handleShareClose = () => setShowShareModal(false);
+
+  const textAreaRef = useRef(null);
+
+  function copyToClipboard(e) {
```
Comment (mecolmg) [inline frontend/src/share-trip/ShareTripModal.js:22]: nit: You don't use e, so no need to define it. Also could be arrow function.

File: frontend/src/share-trip/ShareTripModal.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    textAreaRef.current.select();
+    document.execCommand('copy');
+  }
+
+  return (
+    <Modal show={showShareModal} onHide={handleShareClose}>
+      <Modal.Header closeButton>
+        <Modal.Title className={styles.title}>
+          <FontAwesomeIcon icon={faUmbrellaBeach} className={styles.titleIcon} />
+          {`Share ${tripName}`}
```
Comment (mecolmg) [inline frontend/src/share-trip/ShareTripModal.js:32]: nit: 
```suggestion
          Share {tripName}
```

File: frontend/src/tripUtils.js (hunk 9 above)
Comment (mecolmg) [reply frontend/src/tripUtils.js:17]: I commented on this in the TripCRU PR. The backend should be fixed, and then this should all be removed.

**Pull request merged**